app/models/concerns/alegre_v2.rb

```diff
@@ -263,26 +327,138 @@ def get_confirmed_items(project_media, field)
+      # responses are non-empty before proceeding to creating relationships.
+      Hash[required_keys.collect{|k,v| [k, (Hash[YAML.load(redis.get(v)).collect{|kk,vv| [kk.to_i, vv]}] rescue [])]}]
+    end
+
+    def get_similar_items_v2_callback(project_media, field)
+      type = get_type(project_media)
+      if !should_get_similar_items_of_type?('master', project_media.team_id) || !should_get_similar_items_of_type?(type, project_media.team_id)
+        return {}
+      else
+        cached_data = get_cached_data(get_required_keys(project_media, field))
+        if !cached_data.values.include?(nil)
+          suggested_or_confirmed = cached_data[:suggested_or_confirmed_results]
+          confirmed = cached_data[:confirmed_results]
+          merge_suggested_and_confirmed(suggested_or_confirmed, confirmed, project_media)
+        end
       end
     end
 
     def relate_project_media(project_media, field=nil)
       self.add_relationships(project_media, self.get_similar_items_v2(project_media, field)) unless project_media.is_blank?
     end
 
-    def get_items_with_similar_media_v2(media_url, threshold, team_ids, type)
-      alegre_path = ['audio', 'image'].include?(type) ? self.sync_path_for_type(type) : "/#{type}/similarity/search/"
-      # FIXME: Stop using this method from v1 once all media types are supported by v2
-      # FIXME: Alegre crashes if `media_url` was already requested before, this is why I append a hash
-      self.get_items_with_similar_media("#{media_url}?hash=#{SecureRandom.hex}", threshold, team_ids, alegre_path)
+    def relate_project_media_async(project_media, field=nil)
+      self.get_similar_items_v2_async(project_media, field) unless project_media.is_blank?
+    end
+
+    def relate_project_media_callback(project_media, field=nil)
+      self.add_relationships(project_media, get_similar_items_v2_callback(project_media, field)) unless project_media.is_blank?
+    end
+
+    def get_items_with_similar_media_v2(args={})
+      media_url = args[:media_url]
+      project_media = args[:project_media]
+      threshold = args[:threshold]
+      team_ids = args[:team_ids]
+      type = args[:type]
+      if ['audio', 'image'].include?(type)
+        if project_media.nil?
+          project_media = TemporaryProjectMedia.new
+          project_media.url = media_url
+          project_media.id = Digest::MD5.hexdigest(project_media.url).to_i(16)
+          project_media.team_id = team_ids
+          project_media.type = type
+        end
+        get_similar_items_v2_async(project_media, nil)
+        cached_data = get_cached_data(get_required_keys(project_media, nil))
+        timeout = 60
+        start_time = Time.now
+        while start_time + timeout > Time.now && cached_data.values.include?([])
+          sleep(1)
+          cached_data = get_cached_data(get_required_keys(project_media, nil))
+        end
```

Comment:
@DGaffney can you please notify Sentry if we reach the timeout? This is how you can do it: 

(1) Define a specific class for this error, something like (at the top of the file):

```
class AlegreTimeoutError < StandardError; end
```

(2) Call Sentry:

```
CheckSentry.notify(AlegreTimeoutError.new('Timeout when waiting for async response from Alegre'), params: { project_media: project_media })
```